### ISSUE TITLE
custom schema

### ISSUE THREAD

**Issue by araera111** [araera111/mysql-to-zod#56]
This is a premature comment but I have been thinking about some of this overall ... Rather than using any primitive/defined Zod types like z.string(), z.date(), or z.buffer(), maybe a more universal approach would be to use schema named m2z.BIGINT, m2z.DATETIME, m2z.BOOLEAN, etc. Put the interpretations of these types into a separate file. This way, a developer can use this package to generate the schema structure from MySQL, but the actual interpretation of types isn't hardcoded, and can be modified by developers as they see fit.

I already do this in my code, using "schemaDateTime", "schemaBoolean", etc, so that I don't even link to a specific database. That opens yet another dimension for this package - it doesn't need to be "MySQL-to-Zod, it can be "RDB-to-Zod" or "SQL-to-Zod". A different toZod.ts module can be used for different DB platforms like Postgres, SQL Server, or Oracle.

For your purposes in this discussion, maybe you don't need to worry about the exact implementation, provide one default named m2zBuffer, m2z.buffer, or schemaBuffer. Then later, if you have ideas for different implementations, just add them. The default code can be commented out by developers so that they can use an alternative at runtime. They can even change the implementation for their runtime purpose as they wish (though this probably won't happen) by changing m2zBuffer in their application code to something like m2zBufferV1.

Example:

schemaImplementation.ts

```
export default const m2z = {
   const schemaDate = z.date()
   /** Alternates
   const schemaDate = moreComplexDateSchema()
   const schemaDate = sillyTestDateSchema()
   */
}

schema.ts
import m2z from './schemaImplementation'

export const FooSchema = z.object( {
  some_date : m2z.schemaDate
});
```

I don't think there's a need for even More sophisitication but if someone wanted to make this more complex, all of the schema can be made available, and selected at runtime via option or hardcoded parameter.

```
   const schemaDate = (which) => {
      switch(which) {
        case 0: return z.date; // always default with 0
        case 1: return moreComplexDateSchema
        case 2: return sillyTestDateSchema
      }
   } 
```

As I said, these are premature thoughts, but this is where my mind has been going as I think about solving this question about how we deal with the very real possiblity of multiple valid implementations for a given SQL Type.

This also fits with a couple of my comments about making decisions about implementations with the config.js, and a run-time config mechanism that allows an application that uses code generated here to select an application-specific choice of schema at runtime.

_Originally posted by @TonyGravagno in https://github.com/araera111/mysql-to-zod/discussions/47#discussioncomment-6584831_

**Comment by araera111** [comment]
I am trying.

**Comment by araera111** [comment]
WIP.
I created a customSchema option.
If the TYPE matches, the schemaName is REPLACED.

The following is an idea, although some parts have not yet been implemented.
This is an array of strings of length 3 or 4.
type: [string, string] | [string, string, string] | [string, string, string, string]

[0->TYPE, 1->SchemaName, 2->Import(optional)...path?, 3->Comment(optional)]

I think it would be annoying to write ```import {aaa} from 'bbb'`` every time for import statement, so I would like to only need the bbb part (i.e., just the path).

3->commentは、まだ考えている最中です。
3->comment is still under consideration.

["", "z.number().min(0).max(999)","","min0, max999"]
"comment = "min0, max999" -> comment.includes('min0, max999') -> true -> z.number().min(0).max(999)

If the comment is an includes, the type is ignored and the schemaName is forced to change. This is an idea.
How about this idea? I have not implemented it at all yet.

**Comment by TonyGravagno** [comment]
I am concerned that we lose TypeScript Intellisense when the code is provided as a string. I think if the type is ZodAny[] rather than [string,....] then that issue goes away.
I will look at your code - at this moment I am not sure about how you did or want to implement this...

Thanks!

**Comment by araera111** [comment]
Temporary completion.

This is example config.
```js:mysqlToZod.config.js
const options = {
  outFilePath: "mysqlToZod",
  fileName: "schema.ts",
  customSchema: [
    ["DATETIME", "z.stringg()"],
    [
      "TIMESTAMP",
      "timestampSchema",
      'import { timestampSchema } from "./globalSchema.ts"',
    ],
    [
      "",
      "creationDateSchema",
      'import { creationDateSchema } from "./globalSchema.ts"',
      "CreationDa",
    ],
    [
      "",
      "blogTitleSchema",
      'import { blogTitleSchema } from "./globalSchema.ts"',
      "/^B.*e$/",
    ],
  ],
};
module.exports = options;

```

## DATETIME
```sql
CREATE TABLE `testints` (
  `id` int NOT NULL AUTO_INCREMENT,
  `test1` tinyint(1) NOT NULL,
  `test2` smallint NOT NULL,
  `test3` int NOT NULL,
  `test4` bigint NOT NULL,
  `created` datetime DEFAULT NULL,
  `modified` datetime DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb3
```

```ts
export type ExampleTable = z.infer<typeof example_tableSchema>;

export const testintsSchema = z.object({
  id: z.number(),
  test1: z.number(),
  test2: z.number(),
  test3: z.number(),
  test4: z.number(),
  created: z.string().nullish(), // basic: z.date() but custom z.string()
  modified: z.string().nullish(), // same as above
});
```

## TIMESTAMP

```sql
CREATE TABLE `todo` (
  `id` int NOT NULL AUTO_INCREMENT,
  `task` varchar(255) NOT NULL,
  `due_date` date DEFAULT NULL,
  `created_at` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
  `updated_at` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `comment` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci COMMENT='MY_TODO_LIST
```

```ts
import { timestampSchema } from "./globalSchema.ts";

// [table:todo] : MY_TODO_LIST
export const todoSchema = z.object({
  id: z.number(),
  task: z.string(),
  due_date: z.date().nullish(),
  created_at: timestampSchema.nullish(),
  updated_at: timestampSchema.nullish(),
  comment: z.string().nullish(),
});
export type Todo = z.infer<typeof todoSchema>
```

## regexp or includes

```sql
CREATE TABLE `blog` (
  `title` varchar(30) DEFAULT NULL COMMENT 'BlogTitle',
  `d` date DEFAULT NULL COMMENT 'CreationDate'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

```ts
import {
  blogTitleSchema,
  creationDateSchema,
} from "./globalSchema.ts";

export const blogSchema = z.object({
  title: blogTitleSchema.nullable(), // BlogTitle is matched by "/^B.*e$/"
  d: creationDateSchema.nullable(), //CreationData includes "CreationDa"
});
export type Blog = z.infer<typeof blogSchema>;
```

1. If null is found, nullable is automatically added, but nullable may not be necessary for custom schema.
2. CustomSchema is a string[][], but object is easier to write implementation.
However, when writing configurations, it is easier to use string[][].

## string[][]
```ts
[
  "TIMESTAMP",
  "timestampSchema",
  'import { timestampSchema } from "./globalSchema.ts"',
],
```

## object
```ts
{
  type: "TIMESTAMP",
  schemaName: "timestampSchema",
  importDeclaration: 'import { timestampSchema} from "./globalSchema.ts"'
  comment: undefined
}
```

**Comment by TonyGravagno** [comment]
I will load this today, test, and incorporate your documentation into the updated ReadMe.

**Comment by TonyGravagno** [comment]
The options now have `schema` and `customSchema`. The `schema` and `type` options are closely aligned and until now only detail the generation of the names.

What do you think about moving the new customSchema to schema>zod?

```js

  schema: {
    format: 'camel',
    prefix: '',
    suffix: 'Schema',
    replacements: [],
    nullType: 'nullable',
    zod: [
      ['DATETIME', 'z.string()'],
      ['TIMESTAMP', 'timestampSchema', 'import { timestampSchema } from "./globalSchema.ts"'],
      [
        '',
        'creationDateSchema',
        'import { creationDateSchema } from "./globalSchema.ts"',
        'CreationDate',
      ],
      ['', 'blogTitleSchema', 'import { blogTitleSchema } from "./globalSchema.ts"', '/^B.*e$/'],
    ],
  },
```

**Comment by TonyGravagno** [comment]
In the examples above `./globalSchema.ts` provides the schema. For TIMESTAMP there is no implementation. This implies globalSchema is generated. General format:

```js
import z from 'zod'
export const timestampSchema = z.date()
export const blogTitleSchema = z.string().regex('/^B.*e$/')
```

Do you intend to create a default, like for TIMESTAMP, when an implementation is not provided?

I think it puts too much work on you and this utility to process one of the schema parameters like `'/^B.*e$/'` into `z.string().regex('/^B.*e$/')'`. I think it's completely acceptable for the value to be the entire Zod schema that the user/developer wants :  

```ts
 [
    '',
    'blogTitleSchema',
    'import { blogTitleSchema } from "./globalSchema.ts"',
    "z.string().regex('/^B.*e$/').min(10).max(60)"
  ],
```

<hr>

However (180 degrees opposite) : Since schema can be very complex, I don't think it's necessary for `options` to include it. It will be very difficult for a developer to encapsulate their schema into a quoted string just for this purpose. 

### New globalSchema.ts
I suggest a new `globalSchema.ts` file that has this default:

```ts
import z from 'zod'
const globalSchema = {
  mysqlDate : z.date(),
  mysqlBoolean : z.boolean(),
  mysqlBigInt : z.bigint()
}
module.exports = globalSchema
```

Now schema.ts can start with:

```ts
import { z } from 'zod'
import {globalSchema} from './globalSchema'
```

And for DATE, BOOLEAN, and BIGINT types, use:

```ts
  id :  globalSchema.mysqlBigInt,
  startDate : globalSchema.mysqlDate.nullable(), // nullable based on SQL
```

This does work:
![image](https://github.com/araera111/mysql-to-zod/assets/499473/00dfccb8-2f9d-4288-9dd3-cf33d7b81e7b)

A developer can add anything they want to the globalSchema.

The options>schema>zod>implementation configuration can become another From/To map of SQL types to schema names:

```js
schema: {
  zod: {
    implementation: [
       [ 'DATE' , 'globalSchema.mysqlDate' ],
       [ 'BOOL' , 'globalSchema.mysqlBoolean' ],
       [ 'BOOLEAN' , 'globalSchema.mysqlBoolean' ],
       [ 'BIT' , 'globalSchema.mysqlBoolean' ],
       [ 'BIGINT' , 'globalSchema.mysqlBigInt' ],
       // anything not specified here defaults to direct z.function()
    ]
  }
}
```

This mechanism addresses another concern where a developer might say "I don't like the default implementation of type ZZZ, which might be z.string(). They can easily set in the above options: `ZZZ: 'globalSchema.localZzz'`, and add/modify the globalSchema to their needs. So you don't need to worry about how to craft globally acceptable schema anymore. Note that the user/developer can choose to create their own schema files to use in addition to globalSchema.   
That might result in :  `[ 'TINYINT(1)' , 'appSchema.IntAsBool' ]`  
which might be implemented as `IntAsBool : z.string().length(1).or(z.boolean())`

If you like this implementation, I offer to modify the latest and submit a PR.

**Comment by araera111** [comment]
I appreciate your point of view.
I did not explain myself well enough. My apologies.

We need to think about who will prepare the customSchema.

1. read DB configuration values and prepare default schema
tinyint(1), DATETIME, YEAR...

2. If you need a schema other than the default one, you need to prepare it yourself.
(I did not prepare globalSchema.ts this time due to time constraints)

If you already have a type in your business logic, [2] is sufficient.
However, if you want to make the type a little stricter than the default, it would be convenient to have some types prepared in [1].

Do you think we should create a basic globalSchema based on the DB?

I think my idea about [comment] needs a little more elaboration.
This is to accommodate numbers where the TYPE is INT, but only numbers 0 through 10 are allowed due to comment constraints.

```sql
CREATE TABLE `todo` (
  `id` int NOT NULL AUTO_INCREMENT',
  `task` varchar(255) NOT NULL,
  `test_num` int NOT NULL COMMENT 'this_is_0to10',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci COMMENT='MY_TODO_LIST
```
↓converted

```ts
export const todoSchema = z.object({
  id: z.number(),
  task: z.string(),
  test_num: z.number().min(0).max(10) // comment includes this_is_0to10
});
export type Blog = z.infer<typeof blogSchema>;
```

[TYPE, SchemaName, importDeclaration] If [comment] does not exist, as shown on the left
["INT", "zeroToTenSchema", "import {zeroToTenSchema} from './globalSchema.ts'"]
In this case, all INT types are zeroToTenSchema.

Therefore, a [COMMENT] is required.
["", "zeroToTenSchema", "import {zeroToTenSchema} from './globalSchema.ts'", "0to10"]

In this case, it is not judged by TYPE, but by whether [0to10] is included in the comment.

regexp not for use with z.regexp,
Created for those who want to determine if a comment contains something by regexp rather than includes.


In conclusion, I think it would be better to limit it to [TYPE, schemaPath, comment].

I think it is a good idea to unify import with globalSchema.
I will try to change to that implementation next time.

Also, I would like to stop using quoted strings, as they are quite error-prone (I also lost track of whether they were double or single quotes during testing).

I will continue to make improvements. Thank you for your cooperation.

**Comment by araera111** [comment]
move customSchema to schema.zod

```ts
  schema: {
    format: "camel",
    prefix: "",
    suffix: "Schema",
    replacements: [],
    nullType: "nullish",
    zod: [["DATETIME", "z.string()"]],
  },
```

**Comment by araera111** [comment]
sorry. I move schema.zod to schema.zod.implementation.

```js
const option = {
  schema: {
    etc...,
    zod: {
      implementation: []
    },
  },
}
```

Does this mean that globalSchema.ts will always be created even if schema.zod.implementation is empty?

example:option.schema.implementation.zod.implementation is empty.

```sql
CREATE TABLE `blog` (
  `title` varchar(30) DEFAULT NULL,
  `d` date DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

```ts
import {
  globalSchema,
} from "./globalSchema.ts";

export const blogSchema = z.object({
  title: globalSchema.mysqlString.nullable(), 
  d: globalSchema.mysqlDate.nullable, 
});
export type Blog = z.infer<typeof blogSchema>;
```

Would you prefer it to be this way?

**Comment by TonyGravagno** [comment]
Sincere thanks for all of these exchanges! 🙏 

### About  [COMMENT], 0to10, etc :
With respect, I would not use this feature. I do not see the value in creating a new "language" to describe simple features. Such a comment could easly be ".min(0).max(10)". We would use as many characters to describe the intent, and gain nothing from the use of new and proprietary syntax.

### schema.zod.implementation
Thank you for this. To be clear, the reason I proposed this structure is similar to my reasoning for root-level values like `addTablePrefix:true`. As soon as we define one value, we realize we need more, which then requires deprecating the dedicated value and moving it under a structure.  
With the new `schema` value, everything is nicely under a common structure. I originally thought `schema.zod` too but then thought that there might be more options related to Zod under Schema, so I tried to think of a value name that appropriately represents only the `z.functions()` that implement our intent for the schema ... and that led to `schema.zod.implementation`.

**Comment by TonyGravagno** [comment]
> Does this mean that globalSchema.ts will always be created even if schema.zod.implementation is empty?

I have no firm preference about this, but in terms of elegance I think there is value to moving all schema into a separate file. I did this in my code and I like how it looks. Perhaps with an option like schema>inLine, if true (default), we get the original format:

```js
export const EventSchema = z.object({
  id: z.number(),
  title: z.string(),
  date: z.date(),
```

With inLine=false:

```js
export const EventSchema = z.object({
  id: globalSchema.mysqlNumber,
  title: globalSchema.mysqlString,
  date: globalSchema.mysqlDate,
```

So, yes, with inLine=false, even if there is no custom schema some developers might welcome the freedom to change the implementation of any of these schema later with no application code changes.

This  option determines what gets injected for inLine=true, and for the actual implementation if the schema is in another file:

```js
schema.zod.implementation : {
  [ 'DATE' , 'z.date()' ],
  [ 'DATETIME', 'z.date()'],
  [ 'BOOL' , 'z.boolean()' ],
  [ 'VARCHAR' , 'z.string()' ],
```

This comes from your array that links types to names :

```js
schema.zod.references: { // new object?
  [ 'DATE' , 'mysqlDate' ],
  [ 'DATETIME', 'mysqlDateTime'],
  [ 'BOOL' , 'mysqlBoolean' ],
  [ 'VARCHAR' , 'mysqlString' ],
```

The globalSchema always gets generated based on the above:

```js
// globalSchema.ts:
const globalSchema = {
  mysqlDateTime: 'z.date()' // name from references, value from implementation
}
```

Now, a developer has more options. For example to change DATETIME implementation :

```js
 implementation: {
  ['DATETIME', 'z.string().datetime()'],
...
  references: {
  [ 'DATETIME', 'ourDateTime'],
```

Final generated code:

```js
export const EventSchema = z.object({
  id: globalSchema.mysqlNumber,
  title: globalSchema.mysqlString,
  date: globalSchema.ourDateTime,
```

Oops, schema needs to change for all DATETIME. No problem, change it in globalSchema, with no application changes :  

```js
 implementation: {
  ['DATETIME', 'z.string().datetime({ precision: 3 })'],

**Comment by TonyGravagno** [comment]
For an enhancement ( later, maybe ) ...

As we have both noted earlier, putting code in a string is not ideal.

I have not thought about this much, but a developer might want to set options without quotes, for the benefit of code analysis (IntelliSense, linting, formatting, etc). The toZod code might then read options as a string file to extract the code-as-string to inject into new code.

```js
  [ 'DATE' , z.date() ],
  [ 'DATETIME', z.string().datetime({ precision: 3 }) ],
  [ 'BOOL' , z.boolean() ],
  [ 'VARCHAR' , z.string() ],
```

( Even in markdown that looks better )

Unless you/we can think of an elegant way to allow options to be written without quotes while toZod can effectively scrape and process it into new code, this is only a "nice but not required" feature.

**Comment by TonyGravagno** [comment]
I will say again, in case it will help - I am absolutely capable and willing to make the code changes discussed here too. I sincerely appreciate your effort and enthusiam in accepting and processing these ideas. Just let me know if we get a point where you would prefer that I code in features with a promise of not breaking your defaults. :)

**Comment by TonyGravagno** [comment]
About using options without quotes, I actually already have code that I use to parse the output from mysql-to-zod. You can use this to parse the options. The result is an array of each detail line. :

```ts
// Function to extract details from the source code
function extractDetails(sourceCode) {
    const regex = /export const.*= z\.object\(\{([\s\S]*?)\}\)/
    const match = sourceCode.match(regex)
    return match ? match[1].trim().split('\n') : []
}
```

**Comment by araera111** [comment]
## comment
I thought this might be useful to have yesterday but reconsidered.
I don't think I need it so I will remove it once.
If I need it again, I'll add it.

## inline option.
I will add inline option.

## implementation and references
How about implementation and references together?
(similar to the previous implementation)

```ts
[type, impl, ref]
["DATE", "z.date()", "mysqlDate"]
```
The above will eliminate the need to write "DATE" twice.

I am struggling with what to implement. What features would you most like to see added in this library?　After inlineoption, implementation and references are completed, I will try to implement that feature as a top priority.

**Comment by araera111** [comment]
## commentOption
deleted

## inlineOption
added

## implementation and references
added

```ts
import { z } from "zod";
import { caseUnionSchema, nullTypeUnionSchema } from "./common";

/* 
  ex
  schema.zod.references: {
  [ 'DATE' , 'mysqlDate' ],
  [ 'DATETIME', 'mysqlDateTime'],
  [ 'BOOL' , 'mysqlBoolean' ],
  [ 'VARCHAR' , 'mysqlString' ],
*/
export const schemaZodReferencesSchema = z.tuple([z.string(), z.string()]);
export type SchemaZodReferences = z.infer<typeof schemaZodReferencesSchema>;

/* 
  ex
  schema.zod.implementation : {
  [ 'DATE' , 'z.date()' ],
  [ 'DATETIME', 'z.date()'],
  [ 'BOOL' , 'z.boolean()' ],
  [ 'VARCHAR' , 'z.string()' ],
*/
export const schemaZodImplementationSchema = z.tuple([z.string(), z.string()]);
export type SchemaZodImplementation = z.infer<
  typeof schemaZodImplementationSchema
>;

export const SchemaOptionSchema = z.object({
  format: caseUnionSchema.default("camel"),
  prefix: z.string().default(""),
  suffix: z.string().default("Schema"),
  replacements: z.string().array().array().default([]),
  nullType: nullTypeUnionSchema,
  inline: z.boolean().default(true),
  zod: z
    .object({
      implementation: schemaZodImplementationSchema.array().optional(),
      references: schemaZodReferencesSchema.array().optional(),
    })
    .optional(),
});
export type SchemaOption = z.infer<typeof SchemaOptionSchema>;

```

## option
I have split the option.ts

**Comment by araera111** [comment]
now option

```ts
export const mysqlToZodOptionSchema = z.object({
  outFilePath: z.string().optional().default("./mysqlToZod"),
  fileName: z.string().optional().default("schema.ts"),
  dbConnection: z.any().optional(),
  tableNames: z.string().array().optional().default([]),
  comments: optionCommentsSchema.optional(),
  type: typeOptionSchema.optional(),
  schema: SchemaOptionSchema.optional(),
});
```

**Comment by TonyGravagno** [comment]
I like the new option schema and can make it more detailed soon to explicitly agree with docs. For example, dbConnection can be the specific short format for Knex, or the object format for Knex. For now it works, so no changes are requested. Thanks!

Yes, I believe the older options are obsolete,`nulltype` is there, and `isTypeUpperCamel` is now replaced by schema/type>format>pascal.

**Comment by araera111** [comment]
I would like to take this opportunity to also change the option named outFilePath.
I think it would be more common to name it [outDir], like webpack, etc.
I can't remember why I named it outFilePath at all.

As for dbConnection, I will look into the possibility of using another type instead of z.any.

or

```ts
output: {
  outDir: z.string(),
  fileName: z.string()
}
```

How about this format?
Additional options can be added, for example, separate files for each TABLE.

**Comment by TonyGravagno** [comment]
Agreed. As you've seen, my post-processing code scans for the schema for each table and puts table-specific schema into its own file. It would be nice to have this as an option to do this right from this utility.

Another option ... (We're going crazy on these options, but ... it's all good!)

Someone might want the original table name as the output filename: custorddet.ts
Someone might want the result of the schema processing: CustomerOrderDetailSchema.ts
Or the type processing CustomerOrderDetailType.ts

```ts
output: {
  outDir: z.string(),
  fileName: z.string().optional().default('schema.ts'),
  fileNameFormat: z.enum(["table", "schema","type"]).optional()
}
```

or?

```ts
output: {
  outDir: z.string(),
  fileName: z.union(
     [ z.string().optional().default('schema.ts'),
       z.enum(["table", "schema","type"])
     ]
  )
}
```

( I am not certain those schema are desirable, they are only ideas for general consideration. )
